infra/enrolment-records/repository/src/main/java/com/simprints/infra/enrolment/records/repository/local/RoomEnrolmentRecordQueryBuilder.kt

```diff
@@ -0,0 +1,136 @@
+        val (whereClause, args) = buildWhereClause(query)
+        val specificFormat = query.fingerprintSampleFormat ?: query.faceSampleFormat
+
+        val sql = if (specificFormat != null) {
+            "SELECT COUNT(DISTINCT S.$SUBJECT_ID_COLUMN) FROM $SUBJECT_TABLE_NAME S  INNER JOIN  $TEMPLATE_TABLE_NAME T" +
+                " using(subjectId) $whereClause"
+        } else {
+            "SELECT COUNT(DISTINCT S.$SUBJECT_ID_COLUMN) FROM $SUBJECT_TABLE_NAME S $whereClause"
+        }
+        return SimpleSQLiteQuery(sql, args.toTypedArray())
+    }
+
+    fun buildBiometricTemplatesQuery(
+        query: SubjectQuery,
+        pageSize: Int,
+        lastSeenSubjectId: String? = null,
+    ): SimpleSQLiteQuery {
+        // require format to be set for biometric templates query
+        val format = query.fingerprintSampleFormat ?: query.faceSampleFormat
+        require(format != null) {
+            "Must set format for biometric templates query, use buildSubjectQuery or buildCountQuery instead"
+        }
+        val updatedQuery = query.copy(afterSubjectId = lastSeenSubjectId, sort = true)
+        val (whereClause, args) = buildWhereClause(updatedQuery)
+        val orderByClause = buildOrderByClause(updatedQuery)
+        val sql =
+            """
+            SELECT A.*
+            FROM $TEMPLATE_TABLE_NAME A
+             INNER JOIN (
+                SELECT distinct  S.subjectId
+                FROM $SUBJECT_TABLE_NAME S  INNER JOIN  $TEMPLATE_TABLE_NAME T
+                USING(subjectId)
+                $whereClause
+                $orderByClause
+                LIMIT $pageSize
+            ) B USING(subjectId) where A.format ='$format'
+            """.trimIndent()
+        return SimpleSQLiteQuery(sql, args.toTypedArray())
+    }
+
+    fun buildWhereClause(
+        query: SubjectQuery,
+        subjectAlias: String = "S.", // Default alias for subject table, dot included. Empty string for no alias.
+        templateAlias: String = "T.", // Default alias for template table, dot included. Empty string for no alias.
+    ): Pair<String, List<Any?>> {
+        val clauses = mutableListOf<String>()
+        val args = mutableListOf<Any?>()
+        require(!(query.fingerprintSampleFormat != null && query.faceSampleFormat != null)) {
+            "Cannot set both fingerprintSampleFormat and faceSampleFormat"
+        }
+        // to achieve the highest performance, we should not use OR in the where clause
+        query.subjectId?.let {
+            clauses.add("${subjectAlias}$SUBJECT_ID_COLUMN = ?")
+            args.add(it)
+        }
+        query.subjectIds?.takeIf { it.isNotEmpty() }?.let {
+            clauses.add("${subjectAlias}$SUBJECT_ID_COLUMN IN (${it.joinToString(",") { "?" }})")
+            args.addAll(it)
+        }
+        query.afterSubjectId?.let {
+            clauses.add("${subjectAlias}$SUBJECT_ID_COLUMN > ?")
+            args.add(it)
+        }
```

Comment:
Since `subjectId`, `subjectIds` and `afterSubjectId` are mutually exclusive - do we want to have them in a when to prevent the callers from shooting themselves in the foot?